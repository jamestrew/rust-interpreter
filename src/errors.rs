use thiserror::Error;

use crate::types::Spanned;

pub type Result<T> = std::result::Result<T, Spanned<MonkeyError>>;

#[derive(Error, Debug, Clone)]
pub enum MonkeyError {
    #[error("TokenizerError: {0}")]
    TokenizerError(TokenizerError),

    #[error("ParserError: {0}")]
    ParserError(ParserError),
}

#[derive(Error, Debug, Clone)]
pub enum TokenizerError {
    #[error("Unexpected input {0}. Failed to tokenize.")]
    UnexpectedInput(char),

    #[error("Invalid non UTF-8 string found.")]
    NonUTF8Input,
}

#[derive(Error, Debug, Clone)]
pub enum ParserError {
    #[error("Unexpected token '{0}'.")]
    UnexpectedToken(String),

    #[error("Expected token '{0}' not found.")]
    ExpectedTokenNotFound(String),

    #[error("Unexpected EOF.")]
    UnexpectedEOF,
}

macro_rules! impl_from_for_monkeyerror {
    ($($err:tt),+) => {$(
        impl From<$err> for MonkeyError {
            fn from(value: $err) -> Self {
                MonkeyError::$err(value)
            }
        }
    )+}
}

impl_from_for_monkeyerror!(TokenizerError, ParserError);
