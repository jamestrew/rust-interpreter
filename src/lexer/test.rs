use super::*;
use crate::errors::MonkeyError;

struct T<'a> {
    token: TokenKind,
    slice: &'a str,
}

impl<'a> std::fmt::Debug for T<'a> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "(token: '{}', slice: '{}')", self.token, self.slice)
    }
}

enum K<'a> {
    Token(T<'a>),
    Error(MonkeyError),
}

impl<'a> std::fmt::Debug for K<'a> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Token(t) => write!(f, "Token{:?}", t),
            Self::Error(err) => write!(f, "Error({})", err),
        }
    }
}

fn lex(input: &str) -> Vec<Result<Token>> {
    let lexer = Lexer::new(input);
    lexer.collect()
}

fn debug_print(input: &str, token_results: Vec<Result<Token>>) -> Vec<K> {
    let rows = input.lines().collect::<Vec<_>>();

    let mut ret = Vec::new();
    for result in token_results {
        match result {
            Ok(token) => {
                println!("{:?}", token);
                let row_num = rows[token.row_num()];
                let slice = &row_num[token.span()];
                ret.push(K::Token(T {
                    token: token.take(),
                    slice,
                }));
            }
            Err(err) => ret.push(K::Error(err.take())),
        }
    }
    ret
}

macro_rules! snapshot {
    ($name:tt, $input:expr) => {
        #[test]
        fn $name() {
            let tokens = lex($input);
            insta::with_settings!({
                description => $input,
            }, {
                insta::assert_debug_snapshot!(debug_print($input, tokens));
            })
        }
    };
}

snapshot!(basic, "=");
snapshot!(symbols, "=+(){},;");
snapshot!(assignments1, "let five = 5;");
snapshot!(assignments2, "let ten = 10;");
snapshot!(
    assignments3,
    r"let add = fn(x, y) {
        x + y;
    };"
);
snapshot!(assignments4, "let result = add(five, ten);");

snapshot!(operators, "!-/*5;");
snapshot!(equality1, "5 < 10 > 5;");
snapshot!(equality2, "10 == 10;");
snapshot!(equality3, "10 != 9;");
snapshot!(
    conditional,
    r"if (5 < 10) {
        return true;
    } else {
        return false;
    }"
);

snapshot!(string_literal, r#""hello world";"#);
snapshot!(arrays, "[1, 2]");
snapshot!(dictionary, r#"{ "foo": "bar" };"#);
