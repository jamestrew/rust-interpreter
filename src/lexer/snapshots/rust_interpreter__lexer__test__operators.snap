---
source: src/lexer/test.rs
description: "!-/*5;"
expression: "debug_print(\"!-/*5;\", tokens)"
---
[
    Token(token: '!', slice: '!'),
    Token(token: '-', slice: '-'),
    Token(token: '/', slice: '/'),
    Token(token: '*', slice: '*'),
    Token(token: '5', slice: '5'),
    Token(token: ';', slice: ';'),
    Token(token: 'EOF', slice: ''),
]
